File: 14.js
var _=require('underscore');
function numTerms(i){
	var terms = 1;
	while(i!=1){
		if(i%2)
			i=3*i+1
		else
			i=i/2
		terms++;
	}
	return terms;
}
var maxTerms = start=0;
for(i = 2;i<1000000;i++){
	var num = numTerms(i)
	if(num>maxTerms){
		start=i
		maxTerms = num
	}
}
console.log(start);

File: 14.cpp
#include <iostream>
#define MAX 1000000

unsigned int dict[MAX];

long int collatz_length(long long int start)
{
	long int count=1;
	long long int num=start;
	if(dict[start]!=0)
		return dict[start];
	//Now we calculate the length, if not in dict
	while(num!=1)
	{
		if(num%2==0)
			num=num/2;
		else
			num=3*num+1;
		count++;
	}
	dict[start]=count;
	return count;
}

int main()
{
	long int i=0;
	//set chain length to zero for all numbers
	for(i=0;i<=MAX;i++)
		dict[i]=0;
	//Now we calculate the collatz length for all numbers below 1M
	long int max=0;
	long int maxNum=0;
	for(i=10;i<=MAX;i++)
	{
		long int length=collatz_length(i);
		if(length>max)
		{
			maxNum=i;
			max=length;
		}
	}
	std::cout<<maxNum<<" has a length of "<<max;
	return 0;
}